### PR TITLE
CarShsringApp

### PR DESCRIPTION
**Pull request by OlegShvets1**
CarShsringApp

### CODE REVIEW

File: src/main/java/mate/academy/carsharing/config/MapperConfig.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package mate.academy.carsharing.config;
+
+import org.mapstruct.InjectionStrategy;
+import org.mapstruct.NullValueCheckStrategy;
+import org.mapstruct.NullValuePropertyMappingStrategy;
+
+@org.mapstruct.MapperConfig(
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/config/MapperConfig.java]: import?

File: src/main/java/mate/academy/carsharing/config/OpenApiConfig.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+public class OpenApiConfig {
+    @Bean
+    public OpenAPI customOpenApi() {
+        return new OpenAPI()
+                .components(new Components().addSecuritySchemes("BearerAuth",
+                        new SecurityScheme()
+                                .type(SecurityScheme.Type.HTTP)
+                                .scheme("bearer")
+                                .bearerFormat("JWT")))
+                .addSecurityItem(new SecurityRequirement().addList("BearerAuth"));
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/config/OpenApiConfig.java]: can save to constants

File: src/main/java/mate/academy/carsharing/dto/car/CreateCarRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        @NotBlank
+        String model,
+        @NotBlank
+        String brand,
+        @NotBlank
+        String type,
+        @Min(0)
+        Integer inventory,
+        @Min(0)
+        BigDecimal dailyFee
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/dto/car/CreateCarRequestDto.java]: Positive

File: src/main/java/mate/academy/carsharing/exception/CustomGlobalExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex,
+            HttpHeaders headers,
+            HttpStatusCode status,
+            WebRequest request
+    ) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/exception/CustomGlobalExceptionHandler.java]: add to constnats

File: src/main/java/mate/academy/carsharing/exception/CustomGlobalExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                .map(this::getErrorMessage)
+                .toList();
+        body.put("errors", errors);
+        return new ResponseEntity<>(body, headers, status);
+    }
+
+    private String getErrorMessage(ObjectError e) {
+        if (e instanceof FieldError) {
+            String field = ((FieldError) e).getField();
+            String message = e.getDefaultMessage();
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/exception/CustomGlobalExceptionHandler.java]: cast it in if conditions like in newest version of java

File: src/main/java/mate/academy/carsharing/model/Car.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Column(name = "brand",nullable = false)
+    private String brand;
+
+    @Column(name = "type", nullable = false)
+    @Enumerated(EnumType.STRING)
+    private Type type;
+
+    @Column(name = "inventory", nullable = false)
+    @Min(0)
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/model/Car.java]: @POsitive

File: src/main/java/mate/academy/carsharing/notification/BotRegister.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+public class BotRegister {
+    private final NotificationTelegramBot notificationTelegramBot;
+
+    @PostConstruct
+    public void registerBot() {
+        try {
+            TelegramBotsApi telegramBotsApi = new TelegramBotsApi(DefaultBotSession.class);
+            telegramBotsApi.registerBot(notificationTelegramBot);
+        } catch (TelegramApiException e) {
+            System.err.println("An error occurred while registering the bot: " + e.getMessage());
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/notification/BotRegister.java]: you should log it

File: src/main/java/mate/academy/carsharing/notification/NotificationTelegramBot.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        List<TelegramUserInfo> userInfoList = telegramUserRepository.findByChatIdWithRoles(chatId);
+
+        if (!userInfoList.isEmpty()) {
+            telegramUserRepository.deleteByChatId(chatId);
+        }
+
+        if (update.hasMessage() && update.getMessage().hasText()) {
+            String message = update.getMessage().getText();
+            if (message.equals("/start")) {
+                // Відправляємо стартове повідомлення і додаємо новий запис
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/notification/NotificationTelegramBot.java]: no need comments in code

File: src/main/java/mate/academy/carsharing/notification/NotificationTelegramBot.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            return;
+        }
+
+        SendMessage message = new SendMessage(chatId, value);
+        try {
+            execute(message);
+        } catch (TelegramApiException e) {
+            throw new TelegramBotMessageException("Couldn't send a message: " + value, e);
+        }
+    }
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/notification/NotificationTelegramBot.java]: You can create service and extrazt some logic to this service, can create some handlers for commands

File: src/main/java/mate/academy/carsharing/repository/car/CarRepository.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.util.List;
+import mate.academy.carsharing.model.Car;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/repository/car/CarRepository.java]:
```suggestion
```

File: src/main/java/mate/academy/carsharing/repository/payment/PaymentRepository.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
+package mate.academy.carsharing.repository.payment;
+
+import java.util.List;
+import java.util.Optional;
+import mate.academy.carsharing.model.Payment;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/repository/payment/PaymentRepository.java]:
```suggestion
```

Comment (MentorMateAcademy) [reply src/main/java/mate/academy/carsharing/repository/payment/PaymentRepository.java]: don`t need this anno

File: src/main/java/mate/academy/carsharing/service/notification/TelegramNotificationServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    @Override
+    public void rentalCreationNotification(RentalResponseDto rental) {
+        Long userId = rental.getUserId();
+
+        String message = "* CONGRATULATIONS, THE NEW RENTAL WAS SUCCESSFULLY CREATED *\n"
+                + "\nRental ID - " + rental.getId()
+                + "\nUser ID - " + userId
+                + "\nCar ID - " + rental.getCarId()
+                + "\nRental Date - " + rental.getRentalDate()
+                + "\nReturn Date - " + rental.getRequiredReturnDate();
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/service/notification/TelegramNotificationServiceImpl.java]: use string text block

File: src/main/java/mate/academy/carsharing/service/payment/PaymentServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+    private Payment updatePaymentStatus(Payment.Status status, Long userId) {
+        Optional<Payment> optionalPayment = paymentRepository
+                .findByStatusAndUserId(Payment.Status.PENDING, userId);
+        if (optionalPayment.isPresent()) {
+            Payment payment = optionalPayment.get();
+            payment.setStatus(status);
+            return paymentRepository.save(payment);
+        } else {
+            return null;
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/service/payment/PaymentServiceImpl.java]: return null is not good practice

File: src/main/java/mate/academy/carsharing/service/payment/strategy/FinePaymentService.java [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+public class FinePaymentService implements PaymentService {
+    private static final BigDecimal FINE_MULTIPLIER = BigDecimal.valueOf(2);
+
+    @Override
+    public BigDecimal calculateAmount(BigDecimal dailyFee, long days) {
+
+        return dailyFee.multiply(
+                        BigDecimal.valueOf(days)
+                ).multiply(FINE_MULTIPLIER)
+                .multiply(BigDecimal.valueOf(100));
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/service/payment/strategy/FinePaymentService.java]: better use methods rightPoint move or smth kile this

File: src/main/java/mate/academy/carsharing/service/payment/strategy/FinePaymentService.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    @Override
+    public boolean isApplicable(Payment.Type type) {
+        return type.equals(Payment.Type.FINE);
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/service/payment/strategy/FinePaymentService.java]: enum not compare by equals

File: src/main/java/mate/academy/carsharing/service/stripe/StripeService.java [hunk 16]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.carsharing.service.stripe;
+
+import com.stripe.exception.StripeException;
+import com.stripe.model.checkout.Session;
+import org.springframework.stereotype.Component;
+
+@Component
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/service/stripe/StripeService.java]:
```suggestion
```

File: src/main/java/mate/academy/carsharing/service/telegrambot/TelegramUserService.java [hunk 17]
```diff
@@ -0,0 +1,14 @@
+package mate.academy.carsharing.service.telegrambot;
+
+import mate.academy.carsharing.model.TelegramUserInfo;
+import org.springframework.stereotype.Component;
+
+@Component
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/service/telegrambot/TelegramUserService.java]:
```suggestion
```

File: src/main/java/mate/academy/carsharing/service/user/UserServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+            newUser.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+            newUser.addRole(customerRole);
+            User savedUser = userRepository.save(newUser);
+            return userMapper.mapToDto(savedUser);
+        } else {
+            try {
+                throw new RoleNotFoundException("Role 'CUSTOMER' not found");
+            } catch (RoleNotFoundException e) {
+                throw new RuntimeException(e);
+            }
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/service/user/UserServiceImpl.java]: looks so strange

File: src/main/java/mate/academy/carsharing/validation/FieldMatchValidator.java [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+        try {
+            Field firstField = value.getClass().getDeclaredField(firstFieldName);
+            firstField.setAccessible(true);
+            final Object firstValue = firstField.get(value);
+            Field secondField = value.getClass().getDeclaredField(secondFieldName);
+            secondField.setAccessible(true);
+            final Object secondValue = secondField.get(value);
+            return Objects.equals(firstValue, secondValue);
+        } catch (final Exception e) {
+            throw new RuntimeException(e);
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/carsharing/validation/FieldMatchValidator.java]: where is message?

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/carsharing/config/MapperConfig.java (hunk 1 above)
Comment (OlegShvets1) [reply src/main/java/mate/academy/carsharing/config/MapperConfig.java:7]: +

**Pull request merged**